greenplum-db/gpdb
# Use pg_regress style testing for pg_upgrade tests

Align upgrade testing with regress style tests. This has a few
advantages over cmockery:
  - Low overhead to add test (no recompile, just SQL)
  - Leverage existing parallelization (runtime: 1 minues 50 seconds)
  - Easy to run focused test (e.g. ./test.sh heap_table)
  - Consistency with existing regress tests

We leveraged existing scripts used with cmockery framework. Workflow
defined in test.sh will run a regress schedule (gpdb5_schedule) against
GPDB 5 to create workload to be upgraded, perform upgrade, and run a
regress schedule (gpdb6_schedule) against GPDB 6 to validate the
upgrade.

In order to run the tests you need to setup GPDB 5 and GPDB 6
(directions in README.md). This has not changed from the cmockery
version. Then either `make installcheck` to run the full suite or
`./test.sh <testcase>` to run a single test.

In order to add a new testcase:
  1. Create a sql file along with corresponding expected file to setup
     GPDB 5 cluster. For example:

     sql/mytest.sql:
       ```
       CREATE TABLE t(a int);
       ```
     expected/mytest.out:
       ```
       CREATE TABLE t(a int);
       CREATE
       ```
  2. Update gpdb5_schedule and parallelize as appropriate. Note: check
     tests must be run serially

     gpdb5_schedule
     ```
     test: mytest
     ```
  3. Create a sql file along with corresponding expected file to
     validate upgraded GPDB 6 cluster. Note: test name is expected to be
     same as previous test prefixed with "upgraded_". For example:

     sql/upgraded_mytest.sql:
       ```
       SELECT * FROM t;
       ```
     expected/upgraded_mytest.out:
       ```
       SELECT * FROM t;
       a
       ---
       (0 rows)
       ```
  4. Update gpdb6_schedule and parallelize as appropriate

     gpdb6_schedule
     ```
     test: upgraded_mytest
     ```
